File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        dp = [0]*len(prices) 
        min_price = prices[0] 
        for i in range(len(prices)):
            dp[i] = max(dp[i-1],prices[i]-min_price)
            min_price = min(min_price,prices[i])
            
        return dp[len(prices)-1]

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if(n == 1) {
            return 0;
        }
        
        int maxProfit = 0;
        int mnSoFar = prices[0];
        for(int i = 1; i < n; ++i) {
            maxProfit = max(maxProfit, prices[i] - mnSoFar);
            mnSoFar = min(mnSoFar, prices[i]);
        }
        
        return maxProfit;
        
        
    }
};